metapool/plate.py

```diff
@@ -412,3 +413,287 @@ def _autopool_norm(pool_failures, total_nmol, min_conc, sample_concs,
                      'MiniPico Pooled Volume'] = my_func
 
     return plate_df
+
+
+class PlateReplication:
+    STATUS_EMPTY = 'empty'
+    STATUS_SOURCE = 'source'
+    STATUS_DESTINATION = 'destination'
+
+    row_letters = [chr(ord('A') + x) for x in range(0, 16)]
+
+    # aka ['blue', 'green', 'red', 'yellow']
+    quadrants = ['1', '2', '3', '4']
+
+    def __init__(self, well_column_name):
+        self.map_to_384 = {}
+
+        for quadrant in PlateReplication.quadrants:
+            self.map_to_384[quadrant] = self._get_quadrant(quadrant)
+
+        if well_column_name is None:
+            self.well_column_name = 'Library Well'
+        else:
+            self.well_column_name = well_column_name
+
+        self._reset()
+
+    def _reset(self):
+        # Used for (re)initialization so that make_replicate() calls can
+        # be made multiple times using the same object.
+
+        # All quadrants begin at one as they represent a potential source
+        # quadrant, even if they are empty. If quad 1 is replicated once,
+        # counter '1' will become 2. If quad 2 is replicated three times,
+        # counter '1' will become 4. If a quadrant becomes a destination then
+        # this value will be overwritten. If empty, this value will be unused.
+        self.rep_counters = {'1': 1, '2': 1, '3': 1, '4': 1}
+
+        self.status = {'1': PlateReplication.STATUS_EMPTY,
+                       '2': PlateReplication.STATUS_EMPTY,
+                       '3': PlateReplication.STATUS_EMPTY,
+                       '4': PlateReplication.STATUS_EMPTY}
+
+        # this allows us to store output dataframes for each quadrant,
+        # overwrite them on demand, and save final concatenation for when
+        # we're done.
+        self.data = {'1': None, '2': None, '3': None, '4': None}
```

Comment:
Since there's three valid states a quad can be in (empty, destination, source) we keep track of it separately in self.state. State is checked before we read from/write to self.data so for the most part it's okay. The only time we iterate through self.data keys is when we concatenate the four quadrants at the end. I'll revise this to go by self.state instead, thanks for catching it! Based on discussions with Rodolfo and the earlier code I assumed all four quadrants would be represented.

(I'm not sure I prefer the separation, but the only alternative I can see would be to make self.data values be another dict that stores the state under one key and the data frame under another.)